### PR TITLE
[Orders with Coupons M4] Add discount to product business logic

### LINKED ISSUE #10202: [Orders with Coupons M4] Add discount to product API call
With this task, we implement the API call to the backend to add the product discount, as well as the requirement to save that data locally. We don't implement any UI here, neither the remove discount call.

### PR DESCRIPTION
**Pull request by toupper**
Closes: #10202 

## Description

With this PR we implement the business logic to add a discount to a product. The EditableOrderViewModel creates a new product input with the discount and passes it to the order synchronizer, in a similar way we do for quantities. We therefore refactor the functions and classes involved to accomplish that.
We add a discount to a product by changing the item total (subtotal - discount). This is done in the `ProductInputTransformer`.
Apart from that, I didn't implement any mechanism to avoid higher discounts than the product price itself, as it doesn't make the order fail, and it's already allowed with (negative) fees. We can end up with a negative order total, but that's also allowed in WC Core.

## Testing instructions

1. Go to Orders.
2. Tap to Create a new order.
3. Add a product.
4. Tap on the product row.
5. Tap to add a discount. Tap on Done.
6. Go back to order details. See that the order total shows the discounted value. You can also check that once you create the order in wp-admin (See screenshots)

## Screenshots

Screenshot 2023-07-11 at 14 23 37

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by iamgabrielma** [comment]
I haven't reviewed the code yet, but testing the branch there seems to be odd behaviour if we happen to either add more than one discount to a product, or edit the discount for a product that already has been discounted. As an example:

* Add a product to an order
* Apply a discount (in my case discounted $0.50)
* Go back to the order, and tap on the same product again to add a new discount.
  * Observe that no previous discount is saved or displayed as persisted state, we see "0"
  * Apply a discount again (in my case the second discount was $0.20):
* Observe that the total shows the accumulated discounts (0.50 + 0.20 = 0.70 discount), but the order displays only the latest one, the initial discount shows as lower order cost. The same behaviour can be seen in wp-admin, the total of the order includes the first discount, and only the second discount appears as such:

| Header | Header |
|--------|--------|
|  | Screenshot 2023-07-12 at 12 02 28 |

**Comment by toupper** [comment]
> I haven't reviewed the code yet, but testing the branch there seems to be odd behaviour if we happen to either add more than one discount to a product, or edit the discount for a product that already has been discounted. As an example:

Thanks for the comment @iamgabrielma

I will handle those cases (removal of a discount, edit the discount, adding discount to several products) in a following PR, for this one we should basically just test the scenario described in the PR description: one discount added to one product.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 1]
```diff
@@ -834,14 +842,14 @@ private extension EditableOrderViewModel {
         for product in products {
             // Only perform the operation if the product has not been already added to the existing Order
             if !itemsInOrder.contains(where: { $0.productID == product.productID }) {
-                productInputs.append(OrderSyncProductInput(product: .product(product), quantity: 1))
+                productInputs.append(OrderSyncProductInput(product: .product(product), quantity: 1, discount: 0))
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: We'll be passing a different value as `discount` property in these cases where we initialize it as `0`? If that's not the case, then we could update the `OrderSyncProductInput` to provide a default value:
```diff
struct OrderSyncProductInput {
    enum ProductType {
        case product(Product)
        case productID(Int64)
        case variation(ProductVariation)
    }
    var id: Int64 = .zero
    let product: ProductType
    let quantity: Decimal
-    let discount: Decimal
+    var discount: Decimal = .zero

    func updating(id: Int64) -> OrderSyncProductInput {
        .init(id: id, product: self.product, quantity: self.quantity, discount: discount)
    }
}
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductInOrder.swift [hunk 2]
```diff
@@ -103,7 +99,9 @@ struct ProductInOrder_Previews: PreviewProvider {
                                             canChangeQuantity: false,
                                             imageURL: nil)
         let viewModel = ProductInOrderViewModel(productRowViewModel: productRowVM,
-                                                onRemoveProduct: {})
+                                                baseAmountForDiscountPercentage: 0,
+                                                onRemoveProduct: {},
+                                                onSaveFormattedDiscount: {_ in })
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductInOrder.swift]:
```suggestion
                                                onSaveFormattedDiscount: { _ in })
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift [hunk 3]
```diff
@@ -23,9 +23,10 @@ struct OrderSyncProductInput {
     var id: Int64 = .zero
     let product: ProductType
     let quantity: Decimal
+    let discount: Decimal
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift]: I saw this change after writing the previous comment: Would this change make it easier to declare for general scenarios and tests, where discounts are not involved? 
```suggestion
    var discount: Decimal = .zero
```

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/ProductInputTransformerTests.swift [hunk 4]
```diff
@@ -158,16 +158,38 @@ class ProductInputTransformerTests: XCTestCase {
         XCTAssertEqual(item.total, "19.98")
     }
 
+    func test_sending_an_product_with_a_discount_input_updates_item_on_order() throws {
```
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/ProductInputTransformerTests.swift]:
```suggestion
    func test_sending_a_product_with_a_discount_input_updates_item_on_order() throws {
```

**Review by iamgabrielma — APPROVED** [review]
LGTM! 🚢

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 1 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: Good point! Done in aaa3a1d44cfb7b71280fe877ce99c57bf204fbdb

**Pull request merged**